src/storage/mod.rs

```diff
@@ -392,143 +398,202 @@ impl Options {
+            worker: self.worker.clone(),
+            worker_scheduler: self.worker_scheduler.clone(),
+            read_pool: self.read_pool.clone(),
+            gc_worker: self.gc_worker.clone(),
+            refs: self.refs.clone(),
+            max_key_size: self.max_key_size,
+        }
+    }
+}
+
+impl<E: Engine> Drop for Storage<E> {
+    #[inline]
+    fn drop(&mut self) {
+        let refs = self.refs.fetch_sub(1, atomic::Ordering::SeqCst);
+
+        debug!(
+            "Storage ({} engine) de-referenced (reference count before operation: {})",
+            self.engine, refs
+        );
+
+        if refs != 1 {
+            return;
+        }
+
+        let mut worker = self.worker.lock().unwrap();
+        if let Err(e) = worker.schedule(Msg::Quit) {
+            error!("Failed to ask scheduler to quit: {:?}", e);
+        }
+
+        let h = worker.stop().unwrap();
+        if let Err(e) = h.join() {
+            error!("Failed to join sched_handle: {:?}", e);
+        }
+
+        let r = self.gc_worker.stop();
+        if let Err(e) = r {
+            error!("Failed to stop gc_worker: {:?}", e);
+        }
+
+        info!("Storage ({} engine) stopped.", self.engine);
+    }
+}
+
 impl<E: Engine> Storage<E> {
     pub fn from_engine(
         engine: E,
         config: &Config,
         read_pool: ReadPool<ReadPoolContext>,
+        local_storage: Option<Arc<DB>>,
+        raft_store_router: Option<ServerRaftStoreRouter>,
     ) -> Result<Self> {
-        info!("storage {:?} started.", engine);
-
         let worker = Arc::new(Mutex::new(
             Builder::new("storage-scheduler")
                 .batch_size(CMD_BATCH_SIZE)
                 .pending_capacity(config.scheduler_notify_capacity)
                 .create(),
         ));
         let worker_scheduler = worker.lock().unwrap().scheduler();
-        let gc_worker = GCWorker::new(engine.clone(), config.gc_ratio_threshold);
+        let runner = Scheduler::new(
+            engine.clone(),
+            worker_scheduler.clone(),
+            config.scheduler_concurrency,
+            config.scheduler_worker_pool_size,
+            config.scheduler_pending_write_threshold.0 as usize,
+        );
+        let mut gc_worker = GCWorker::new(
+            engine.clone(),
+            local_storage,
+            raft_store_router,
+            config.gc_ratio_threshold,
+        );
+
+        worker.lock().unwrap().start(runner)?;
+        gc_worker.start()?;
+
+        info!("Storage ({} engine) started.", engine);
+
         Ok(Storage {
             engine,
             worker,
             worker_scheduler,
             read_pool,
             gc_worker,
+            refs: Arc::new(atomic::AtomicUsize::new(1)),
             max_key_size: config.max_key_size,
         })
     }
 
-    pub fn mut_gc_worker(&mut self) -> &mut GCWorker<E> {
-        &mut self.gc_worker
-    }
-
-    pub fn start(&mut self, config: &Config) -> Result<()> {
```

Comment:
I think it's better to keep this function but remove the `pub` and call it internally.

Reply:
In this way, it cannot avoid developers calling it inside Storage (which will cause undefined behaviours).